src/dune_engine/load_rules.ml

```diff
@@ -39,20 +38,32 @@ module Loaded = struct
   let no_rules ~allowed_subdirs =
     Build
       { allowed_subdirs
-      ; rules_produced = Rules.empty
       ; rules_here = no_rules_here
       ; aliases = Alias.Name.Map.empty
       }
 end
 
 module Dir_triage = struct
+  module Build_directory = struct
+    (* invariant: [dir = context_or_install / sub_dir] *)
+    type t =
```

Comment:
The invariant seems to suggest that this should really be a sum type, i.e.:
```
type t =
  | Context_or_install of { dir : Path.Build.t ; context_or_install : Context_or_install.t }
  | Sub_dir of { dir : Path.Build.t ; sub_dir : Path.Source.t }
```
or something like this (perhaps with `dir` pulled to the top). Or does this make the implementation harder?

Reply:
Oh, sorry, I misinterpreted the invariant. Ignore this.